### PR TITLE
Add proposals

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes: 
https://github.com/ballerina-platform/ballerina-standard-library/issues/2689
https://github.com/ballerina-platform/ballerina-standard-library/issues/2772
## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: docs/proposals/extract_substring.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# Returns the first substring in str that matches the regex.
+# ```ballerina
+# regex:Match? result = regex:search("Betty Botter bought some butter but she said the butter’s bitter.", "\\b[bB].tt[a-z]*");
+# ```
+#
+# + str - The string to be matched
+# + regex - The regex value
+# + startIndex - The starting index for the search
+# + return - a `Match` record which holds the matched substring, or nil if there is no match
+public isolated function search(string str, string regex, int startIndex = 0) returns Match? {}
```
Comment (MadhukaHarith92) [inline docs/proposals/extract_substring.md:58]:
```suggestion
public isolated function search(string str, string regex, int startIndex = 0) returns Match?;
```

Comment (MadhukaHarith92) [reply docs/proposals/extract_substring.md:58]: Normally, we use semicolons at the end of an API definition.

File: docs/proposals/extract_substring.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+# Returns all substrings in str that match the regex.
+# ```ballerina
+# regex:Match[] result = regex:searchAll(""Betty Botter bought some butter but she said the butter’s bitter.", "\\b[bB].tt[a-z]*");
+# ```
+#
+# + str - The string to be matched
+# + regex - The regex value
+# + return - An array of `Match` records
+# Each member holds a matched substring
+public isolated function searchAll(string str, string regex) returns Match[] {}
```
Comment (MadhukaHarith92) [inline docs/proposals/extract_substring.md:71]:
```suggestion
public isolated function searchAll(string str, string regex) returns Match[];
```

**Comment by daneshk** [comment]
@kalaiyarasiganeshalingam We need to follow the same format as others. Can you check

**Comment by kalaiyarasiganeshalingam** [comment]
> @kalaiyarasiganeshalingam We need to follow the same format as others. Can you check

Added the missing info now.

File: docs/proposals/extract_substring.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+_Created_: 2022/03/11   
+_Updated_: 2022/03/11  
+_Issues_: [#2689](https://github.com/ballerina-platform/ballerina-standard-library/issues/2689) 
+
+## Summary
+The Ballerina Regex module doesn't have any API to extract a substring/s from the given string by using regex. So, this proposal introduces new APIs to extract a substring/s from the given string using regex.
+
+## Goals
+
+- Provide a way to extract a substring/s using regex.
```
Comment (daneshk) [inline docs/proposals/extract_substring.md:14]:
```suggestion
Provide a way to extract a substring/s using regex.
```

File: docs/proposals/extract_substring.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+# + matched - Matched substring
+# + startIndex - The start index of the match
+# + endIndex - The last index of the match
+type PartMatch record {|
+string matched;
+int startIndex;
+int endIndex;
+|};
+```
+```ballerina
```
Comment (daneshk) [inline docs/proposals/extract_substring.md:33]:
```suggestion

```ballerina
```

File: docs/proposals/update_replace_functions.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    type ReplacerFunction function(Match 'match) returns string;
+    ```
+
+- Introduce a new type to get the replacement value.  
+    ```ballerina
+    # A function to be invoked to create the new substring or string value to be used to replace the matches.
+    public type Replacement ReplacerFunction|string;
+    ```
+
+- Change the 3rd param of the API with a new type.
```
Comment (MadhukaHarith92) [inline docs/proposals/update_replace_functions.md:40]: Instead, shall we say "Change the type of `replacement` param from `string `to `Replacement` type."

File: docs/proposals/update_replace_functions.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+## Summary
+The existing replacement functions don't support the following functionalities:
+- There is no way to replace the matches from the particular string index of the given string.
+- There isn't any mechanism to replace the matches with a dynamic value.
+
+So, this proposal improves the replacement functions to support the above functionalities.
+
+## Goals
+
+- Provide a way to support the above functionalities.
```
Comment (daneshk) [inline docs/proposals/update_replace_functions.md:18]: @kalaiyarasiganeshalingam shall we make new lines and space consistent across the proposal.

File: docs/proposals/extract_substring.md (hunk 4 above)
Comment (kalaiyarasiganeshalingam) [reply docs/proposals/extract_substring.md:33]: Fixed now.

File: docs/proposals/update_replace_functions.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    # + originalString - The original string to replace the first occurrence of the
+    #                    substring that matches the provided regex
+    # + regex - The regex to match the first substring in the `originalString` to
+    #           be replaced
+    # + replacement - The replacement string or function to replace the first substring, which
+    #                 matches the regex
+    # + startIndex - The starting index for the search
+    # + return - The resultant string with the replaced substring
+    public isolated function replace(string originalString, string regex, Replacement replacement, int startIndex = 0) 
+                             returns string;
```
Comment (daneshk) [inline docs/proposals/update_replace_functions.md:72]: This is a new function, right? we need to mention about the deprecating function as well

Comment (kalaiyarasiganeshalingam) [reply docs/proposals/update_replace_functions.md:72]: Yes

Comment (kalaiyarasiganeshalingam) [reply docs/proposals/update_replace_functions.md:73]: Fixed in [7ccd7b2](https://github.com/ballerina-platform/module-ballerina-regex/pull/376/commits/7ccd7b28be99b7c7dc4e285795c03b6400795ed4)

**Review by daneshk — APPROVED** [review]

**Review by MadhukaHarith92 — APPROVED** [review]

**Pull request merged**